### DISCUSSION TITLE
How to handle `nread == 0` on `uv_read_cb`?

### DISCUSSION THREAD

**Discussion by fiatjaf** [libuv/libuv#3729] [Q&A] [answered]
There is a note in the [docs](http://docs.libuv.org/en/v1.x/stream.html?highlight=uv_read_start#c.uv_read_cb) that say:

> nread might be 0, which does not indicate an error or EOF. This is equivalent to EAGAIN or EWOULDBLOCK under read(2).

Forgive me since I am a newbie in these C and low-level matters, but I understood that this means the socket would block, but it won't since it is marked to be read in a nonblocking way. What do I do with this information though?

Apparently `uv_read_cb` is called again after a while, so I just do not do anything when I get `nread == 0`, but sometimes it takes a long time for `uv_read_cb` to be called again. Is that the expected behavior?

If anyone is curious, this is the code I have (creating some very thin bindings for Scala Native to be able to read from local sockets): https://github.com/fiatjaf/sn-unixsocket/blob/f6278dbff079c952a21370e91a0b266f3f3522d9/src/main/scala/UnixSocket.scala#L150-L207

**Comment by bnoordhuis** [comment]
It's a hint from libuv that the memory you allocated in your `uv_alloc_cb` can be released again. If libuv didn't call your `uv_read_cb` with `nread == 0`, you wouldn't know when it's safe to free/reuse.

**Reply by fiatjaf** [reply]
Thank you, but that confuses me even more. I thought the memory was safe to free/reuse when it was when there was an EOF.

**Comment by fiatjaf** [comment, marked as answer]
I think the bug I was seeing was actually a false-flag. This thread may be unnecessary and my implementation may be good in fact. I am not sure.
